extensions/common/json_util.h

```diff
@@ -25,46 +25,88 @@
 namespace Wasm {
 namespace Common {
 
-// Parse JSON. Returns the discarded value if fails.
-::nlohmann::json JsonParse(absl::string_view str);
+using JsonObject = ::nlohmann::json;
+
+enum JsonParserResultDetail {
+  EMPTY,
+  OK,
+  OUT_OF_RANGE,
+  TYPE_ERROR,
+  PARSE_ERROR,
+  INVALID_VALUE,
+};
+
+std::pair<JsonObject, JsonParserResultDetail> JsonParse(absl::string_view str);
 
 template <typename T>
-absl::optional<T> JsonValueAs(const ::nlohmann::json& j);
+std::pair<absl::optional<T>, JsonParserResultDetail> JsonValueAs(
+    const JsonObject&) {
+  static_assert(true, "Unsupported Type");
+}
 
 template <>
-absl::optional<absl::string_view> JsonValueAs<absl::string_view>(
-    const ::nlohmann::json& j);
+std::pair<absl::optional<absl::string_view>, JsonParserResultDetail>
+JsonValueAs<absl::string_view>(const JsonObject& j);
 
 template <>
-absl::optional<std::string> JsonValueAs<std::string>(const ::nlohmann::json& j);
+std::pair<absl::optional<std::string>, JsonParserResultDetail>
+JsonValueAs<std::string>(const JsonObject& j);
 
 template <>
-absl::optional<int64_t> JsonValueAs<int64_t>(const ::nlohmann::json& j);
+std::pair<absl::optional<int64_t>, JsonParserResultDetail> JsonValueAs<int64_t>(
+    const JsonObject& j);
 
 template <>
-absl::optional<bool> JsonValueAs<bool>(const ::nlohmann::json& j);
+std::pair<absl::optional<uint64_t>, JsonParserResultDetail>
+JsonValueAs<uint64_t>(const JsonObject& j);
 
-template <typename T>
-absl::optional<T> JsonGetField(const ::nlohmann::json& j,
-                               absl::string_view field) {
+template <>
+std::pair<absl::optional<bool>, JsonParserResultDetail> JsonValueAs<bool>(
+    const JsonObject& j);
+
+template <class T>
+class JsonGetField {
+ public:
+  JsonGetField(const JsonObject& j, absl::string_view field);
+  const JsonParserResultDetail& detail() { return detail_; }
+  T fetch() { return object_; }
```

Comment:
Can we stick with `value` and `value_or`?

Reply:
done